src/vsr/replica.zig

```diff
@@ -6777,6 +7396,387 @@ pub fn ReplicaType(
+
+        fn sync_cancel_commit_callback(self: *Self) void {
+            assert(!self.solo());
+            assert(self.sync_stage == .cancel_commit);
+            assert(!self.grid_repair_message_timeout.ticking);
+
+            switch (self.commit_stage) {
+                .idle,
+                .next,
+                .next_pipeline,
+                .cleanup,
+                .prefetch_state_machine,
+                .compact_state_machine,
+                .checkpoint_state_machine,
+                => unreachable,
+
+                .next_journal,
+                .setup_client_replies,
+                .checkpoint_client_replies,
+                .checkpoint_superblock,
+                => {},
+            }
+
+            // Even though the commit flow has been stopped, there may still be pending IO
+            // between beats.
+            // TODO Compaction pacing: If there is no grid IO between beats, we can transition
+            // to stage=request_target.
+            self.sync_dispatch(.cancel_grid);
+        }
+
+        fn sync_cancel_grid_callback(grid: *Grid) void {
+            const self = @fieldParentPtr(Self, "grid", grid);
+            assert(self.sync_stage == .cancel_grid);
+            assert(self.grid_writes.executing() == 0);
+            assert(self.grid.read_faulty_queue.count == 0);
+
+            if (self.commit_stage == .idle) {
+                assert(self.commit_prepare == null);
+            } else {
+                if (self.commit_prepare) |prepare| self.message_bus.unref(prepare);
+                self.commit_prepare = null;
+                self.commit_stage = .idle;
+            }
+
+            self.grid_repair_message_timeout.stop();
+            self.sync_dispatch(.request_target);
+        }
+
+        fn sync_request_trailers_callback(self: *Self) void {
+            assert(!self.solo());
+            assert(self.sync_stage == .request_trailers);
+            assert(!self.superblock.updating(.checkpoint));
+            assert(self.superblock.storage.next_tick_queue_lsm.empty());
+            assert(self.commit_stage == .idle);
+            assert(self.commit_prepare == null);
+            assert(!self.grid_repair_message_timeout.ticking);
+            assert(self.grid.read_pending_queue.empty());
+            assert(self.grid.read_faulty_queue.empty());
+            assert(self.grid.write_queue.empty());
+            if (self.status == .normal) assert(!self.primary());
+
+            const stage = self.sync_stage.request_trailers;
+            assert(stage.done());
+
+            // TODO(Zig) Use named format specifiers to avoid mixups.
+            log.debug("{}: sync_request_trailers_callback: " ++
+                "checkpoint_op={} checkpoint_id={x:0>32} " ++
+                "manifest_checksum={x:0>32} " ++
+                "free_set_checksum={x:0>32} " ++
+                "client_sessions_checksum={x:0>32}", .{
+                self.replica,
+                stage.target.checkpoint_op,
```

Comment:
Does it matter where _exactly_ we call `.cache_mark()`? Intuitively it seems we won't be touching grid until we do state_machine.open in `sync_request_manifest_logs`, so perhaps we should call it there? 

I guess I still don't exactly understand the marking stuff. The problem is that, if we don't do marking, aftre sync our cache would've contain blocks which no longer logically exist. This is _performance_ issue (useless blocks occupy the cache) but not a correctness issue (contents of cache and disk is still in sync, it's just we know that it'll be overitten soo). Still, we want to remove these blocks. We _could_ just clear the cache completely, but that would be slow. So what we do is that we track which blocks we touched during sync (including state-machine open), and, when the sync finishes, clear everything else. 

I have two questions here (besides the genreal one of whethere this is remotely correct): 

* why we _removing_ blocks from the cache, rather than just demoting them?
* wouldn't removal lead to extra reads? Let's say we removed block 123, because it probably was overwritten during the part of the history state sync skips over. We don't _actively_ replace it though, so, when the time comes for us to read (the new contents of) 123, we'll do a disk read, notice checksum mismatch, and start repair. But we _had_ a block with the wrong checksum in the cache before, so we could have noticed that we need to repair without fetching the block from disk again. I guess this is just to keep thigs simple? We want both "what's in cache is on disk" and "what's in cache is in the checkpoint we think we are in" invariants to hold?